src/contract.ts

```diff
@@ -1,60 +1,63 @@
-import {
-    ABI,
-    ABISerializableObject,
-    Action,
-    Name,
-    NameType,
-    Session,
-    TransactResult,
-} from '@wharfkit/session'
-import type {APIClient} from '@wharfkit/session'
+import {ABI, ABIDef, APIClient, Name, NameType, Session} from '@wharfkit/session'
+
 import {Table} from './contract/table'
 
+export interface ContractArgs {
+    abi: ABIDef
+    account: NameType
+    client: APIClient
+}
+
 export interface ContractOptions {
-    name: NameType
-    client?: APIClient
-    abi?: ABI.Def
+    session?: Session
 }
 
 /**
- * Represents a smart contract in the blockchain.
+ * Represents a smart contract deployed to a specific blockchain.
  * Provides methods for interacting with the contract such as
  * calling actions, reading tables, and getting the ABI of the contract.
  */
 export class Contract {
-    private static _shared: Contract | null = null
-    private static account: Name
-
-    private abi?: ABI.Def
-
+    readonly abi: ABI
     readonly account: Name
-    readonly client?: APIClient
+    readonly client: APIClient
+    readonly session?: Session
 
     /**
      * Constructs a new `Contract` instance.
      *
+     * @param {ContractArgs} args - The required arguments for a contract.
      * @param {ContractOptions} options - The options for the contract.
-     * @param {NameType} options.name - The name of the contract.
-     * @param {APIClient} options.client - The client to connect to the blockchain.
      */
-    constructor(options: ContractOptions) {
-        this.account = Name.from(options.name)
+    constructor(args: ContractArgs, options: ContractOptions = {}) {
+        this.abi = ABI.from(args.abi)
+        this.account = Name.from(args.account)
+        this.client = args.client
 
-        this.client = options.client
+        if (options.session) {
+            this.session = options.session
+        }
+    }
 
-        this.abi = options.abi
+    static from(args: ContractArgs, options: ContractOptions = {}): Contract {
+        return new this(args, options)
     }
 
-    /**
-     * Creates a new `Contract` instance with the given options.
-     *
-     * @param {ContractOptions} options - The options for the contract.
-     * @return {Contract} A new contract instance.
-     */
-    static from(options: ContractOptions): Contract {
-        return new this(options)
+    get tables(): string[] {
```

Comment:
Personally, I would expect `tables` to return an array of table objects since `table()` returns a single Table object. If you want a method that returns the list of table names, maybe we rename it to something like `tableNames()`?

Reply:
Yeah, we can rename it 👍 

I originally had it where `contract.tables` was populated with all the tables, which gave the syntax of:

```
contract.tables.producers.where()
```

But there was no auto completion on the `contract.tables` property in the "dynamic" mode, since the object had no idea what properties existed on `contract.tables` (they are set in the constructor). 

If we did this pattern, the codegen version would be able to autocomplete the table names, and functionality would have diverged from dynamic vs codegen. If we want that, I'm all for it - but left it out since it was a place where they would have diverged.